### PR TITLE
Fix -1 for BigInt factorial - Update gmp.jl

### LINKED ISSUE #51488: factorial(big(-1)) falsely evaluates to 0 instead of raising error
I noticed the following behaviour when computing the ```factorial``` of negative integers (e.g., ```-1```): While ```factorial(-1)``` raises a domain error (as also python would do), computing this in high precision by ```factorial(big(-1))``` evaluates to ```0```. 

I previously posted this as a question on discourse; please see [there](https://discourse.julialang.org/t/factorial-big-1-0-bug-or-expected-behaviour/104298) for more. The current implementation seems intentional, but it seems like a bug and ```factorial(big(x))``` for negative integer ```x``` should instead raise the same domain error.

```
julia> factorial(-1)
ERROR: DomainError with -1:
`n` must not be negative.
Stacktrace:
 [1] factorial_lookup
   @ ./combinatorics.jl:18 [inlined]
 [2] factorial(n::Int64)
   @ Base ./combinatorics.jl:27
 [3] top-level scope
   @ REPL[4]:1

julia> factorial(big(-1))
0

julia> factorial(big(-5))
0
```

```
julia> versioninfo()
Julia Version 1.8.5
Commit 17cfb8e65ea (2023-01-08 06:45 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin21.4.0)
  CPU: 20 × Intel(R) Core(TM) i9-10910 CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 10 on 20 virtual cores
Environment:
  JULIA_EDITOR = code
  JULIA_NUM_THREADS = 10
  DYLD_FALLBACK_LIBRARY_PATH = [...]
```

### PR DESCRIPTION
**Pull request by PallHaraldsson**
Fixes #51488

### CODE REVIEW

**Comment by PallHaraldsson** [comment]
I noticed also this line can be eliminated: https://github.com/JuliaLang/julia/blob/66fe51f0ad9531fa782c4dece0669076456f396c/base/combinatorics.jl#L20

i.e. 2 checks (branches) or even 1 (not 3), by adding 1 (or zero-based indexing; that will in effect happen, +1 optimized out), if thought to be worth it for one more entry in the lookup table.

**Comment by jmkuhn** [comment]
Add test for `factorial(BigInt(-1))` at https://github.com/JuliaLang/julia/blob/66fe51f0ad9531fa782c4dece0669076456f396c/test/mpfr.jl#L695-L704

**Comment by stevengj** [comment]
LGTM, other than the need for a `@test_throws DomainError factorial(BigInt(-1))`.

**Comment by PallHaraldsson** [comment]
@jmkuhn those tests are maybe in the wrong place. I added at test/gmp.jl, since I kind of want MPFR gone. Actually both.

https://github.com/JeffreySarnoff/ArbNumerics.jl/issues/63#issuecomment-1737941768

**Review by stevengj — APPROVED** [review]

**Comment by KshitijThareja** [comment]
Is this issue still open? I also had the same approach, changing the factorial function in base/gmp.jl and the writing tests for it in test/gmp.jl. Its passed all checks when I tested it locally. Should I make a PR for this issue?

File: test/gmp.jl [hunk 1]
```diff
@@ -220,6 +220,7 @@ end
 end
 @testset "combinatorics" begin
     @test factorial(BigInt(40)) == parse(BigInt,"815915283247897734345611269596115894272000000000")
+    @test_throws DomainError factorial(BigInt(-1))
```
Comment (rfourquet) [inline test/gmp.jl:223]: We might as well test for more negative values:
```suggestion
    @test_throws DomainError factorial(BigInt(-1))
    @test_throws DomainError factorial(BigInt(rand(-999:-1)))
```

**Comment by rfourquet** [comment]
> Should I make a PR for this issue?

Thanks for offering your help @KshitijThareja , but in this case this PR is already open for solving the issue, and we should rather move forward with it.

**Comment by longemen3000** [comment]
bump? test fails seem unrelated

**Pull request merged**